### PR TITLE
feat(schema): make pojo paths optionally become subdoc instead of Mixed (closes #7494)

### LINKED ISSUE #7181: Make `foo: { type: { name: String } }` behave the same as `{ foo: new Schema({ name: String }) }`
Re: #7143. Right now the below declarations are different, and its a very subtle footgun that's hard to notice. 

```javascript
// `foo.name` is a string
new Schema({ foo: { name: String } });

// `foo` is mixed
new Schema({ foo: { type: { name: String } } });
```

Demo using Node shell:

```
$ node
> const { Schema } = require('mongoose')
undefined
> new Schema({ foo: { type: { name: String } } }).path('foo')
Mixed {
  path: 'foo',
  instance: 'Mixed',
  validators: [],
  setters: [],
  getters: [],
  options: { type: { name: [Function: String] } },
  _index: null,
  [Symbol(mongoose:schema_mixed)]: true }
> new Schema({ foo: { type: { name: String } } }).path('foo.name')
undefined
> 
> 
> new Schema({ foo: { name: String } }).path('foo')
undefined
> new Schema({ foo: { name: String } }).path('foo.name')
SchemaString {
  enumValues: [],
  regExp: null,
  path: 'foo.name',
  instance: 'String',
  validators: [],
  setters: [],
  getters: [],
  options: { type: [Function: String] },
  _index: null }
> 
```

Anybody have thoughts on this? I'm thinking it makes sense to deprecate this behavior, and make it so that `new Schema({ foo: { type: { name: String } } })` makes `foo.name` a string in Mongoose 6.x.

@lineus  @fonger ?

### LINKED ISSUE #7408: Schema.path('path to object') returns undefined
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

`MySchema.path('path to object')` returns `undefined`.

**If the current behavior is a bug, please provide the steps to reproduce.**

Having:
```ts
const UserSchema = new Schema({
  test1: {
    a: String
  },
  test2: {
    type: {
      a: String
    }
  },
  test3: [{
    a: String
  }]
});
```
Will make
`MySchema.path('test1')` == `undefined`.

While the other ones work:
`MySchema.path('test2')` != `undefined`.
`MySchema.path('test3')` != `undefined`.

**What is the expected behavior?**

I would expect `test1` to work the same way as `test2`.

**Please mention your node.js, mongoose and MongoDB version.**
Node v11.6.0
Mongoose: 5.4.3
MongoDB: v4.0.4

### LINKED ISSUE #7494: feature request: nested schemas - fully declarative pojo schema definitions
**Do you want to request a *feature* or report a *bug*?**

*feature*

This is very related to #7181 but after starting to write there I decided it's probably off-topic.

TL;DR: 
- hail mary request: nested objects in schema configuration objects make child schemas by default - just like arrays already do.
- next best thing request: an boolean OPTION to do it without needing to use { type: new Schema(...) }

**What is the current behavior?**

- If you want validation on multiple levels (and don't want Mixed), you need a child schema.
- When a schema configuration has an array, a child schema is created automatically, and a simple pojo can fully describe your schema (including subschemas) \o/.

But when it's not an array, it's impossible to do so without:

child: { type: new Schema({...}) }

That means there are certain schemas that can't be fully described with passing a single configuration pojo to Schema() (namely when you need single subdocuments).

"What's the big deal?  Just wrap the child in 'new Schema()' and move on", you say?

I share my schema definitions between /packages/client and /packages/server, and I currently can't have just a simple pojo in /packages/shared.

Instead of this:

``` javascript
export const schema = {
  sub: {
    type: {
      from: Number,
      to: Number
    },
    validate: ...
  }
}

```

I have to turn it into a factory of sorts so I can use the right "mongoose" - something like this:

``` javascript
export const schemaFactory = (mongoose) => ({
  sub: {
    type: new mongoose.Schema({
      from: Number,
      to: Number
    }),
    validate: ...
  }
});

```

**What is the expected behavior?**

Making the two equivalent would be great, but I don't think everyone would want it.  So if instead, we can have a subSchema: true option or something like that it would be awesome possum.

**Please mention your node.js, mongoose and MongoDB version.**

### PR DESCRIPTION
**Pull request by captaincaius**
Note: Might also close #7181 and close #7408 too...

- default behaves as previous releases did
- embeddeddocument test copied to cover declarative syntax

- otherwise, type: somePojo turns somePojo into subdocument and Mixed paths are only created when explicitly requested via Mixed or {}

I think it's probably also a good idea to make this option a base option too, but wanted to ask for blessing first.

Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory.

If you're making a change to documentation, do **not** modify a `.html` file directly. Instead find the corresponding `.pug` file or test case in the `test/docs` directory.

**Summary**

It creates an option that makes type keys valued as POJOs create a subschema instead of Mixed.  The existing behavior is unaltered by default.

This also helps full-stack monorepo use-cases that share the schema configuration object between the client-side project and the server-side project, because the schema configuration object can specify subschemas without importing Mongoose.Schema (so client-side can use mongoose/browser and server-side can use mongoose itself).

**Examples**

See test ;).

### CODE REVIEW

**Comment by captaincaius** [comment]
Please let me know if:

1. This is the right approach (so I can update docs too).
2. You're okay with adding a base option for this too.

**Comment by captaincaius** [comment]
Is it just me or are tests flaky now? :(

File: test/types.embeddeddocumentdeclarative.test.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    };
+
+    ChildSchema = {
+      name: String,
+      children: [GrandChildSchema]
+    };
+
+    ParentSchema = new Schema({
+      name: String,
+      typePojoToMixed: false,
```
Comment (vkarpov15) [inline test/types.embeddeddocumentdeclarative.test.js]: I really don't like this syntax - mixing options with paths is very messy.

Comment (captaincaius) [reply test/types.embeddeddocumentdeclarative.test.js]: Good catch - yeah this was a total typo...  it was supposed to be in the options object.

the fact that this test passes tells me my lazy copy-the-other-test approach isn't sufficient for this.  I'll write tests from scratch and ensure there's no regression when the option isn't set (a Mixed path is made) and that it truly creates subdocuments when the option IS set.

Comment (captaincaius) [reply test/types.embeddeddocumentdeclarative.test.js]: Fixed.

The fact that this test passed scared the life out of me so I rewrote the test entirely - from a coverage POV it's much better.  But from a test independence POV there may be a little redundancy.  I think that if typePojoToMixed ever becomes true by default, a little test refactor might be in order.

**Comment by captaincaius** [comment]
test fixed...  still pending:
- updated docs
- global default for the option (so it doesn't need to be set on every Schema)

**Comment by captaincaius** [comment]
mongoosewide option added - only docs left...  I think I'll plan to update the following:

- subdocument guide: mention the ability to do it without new Schema() with the new option
- SchemaTypes guide: Mixed section

...and looks like that's it AFAICT

**Comment by captaincaius** [comment]
all done - ready for review.
